Ini scanner mode

jcsanyi opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description
Adds a `typedMode` flag to the INI Reader to support the built-in `INI_SCANNER_TYPED` flag on `parse_ini_file()`.
This allows integer, boolean, and null values to be returned as the appropriate types, rather than all being returned as strings.

jcsanyi committed:
Add support for INI_SCANNER_TYPED

Signed-off-by: Jonathan Csanyi <jdc@csanyi.ca>

Ocramius commented on src/Reader/Ini.php:
```diff
@@ -104,6 +113,44 @@ public function getProcessSections()
+     * @param bool $typedMode
+     * @return $this
+     */
+    public function setTypedMode($typedMode)
```
This should just accept `bool` as an input (no cast in method)

Ocramius replied:
Return type is to be specified

Ocramius commented on src/Reader/Ini.php:
```diff
@@ -104,6 +113,44 @@ public function getProcessSections()
+     * @see https://www.php.net/parse_ini_file
+     * @return bool
+     */
+    public function getTypedMode()
```
Return type to be specified

Ocramius commented on src/Reader/Ini.php:
```diff
@@ -104,6 +113,44 @@ public function getProcessSections()
+     * @see https://www.php.net/parse_ini_file
+     * @return int
+     */
+    public function getScannerMode()
```
Return type to be specified

Ocramius reviewed (changes requested):
Minor changes needed, bit otherwise seems OK.

My biggest issue with the patch is that, in order to follow consistency with pre-existing code, setters/getters are used to enable this behaviour, instead of relying on the `Reader\Ini` constructor.

jcsanyi committed:
Add type hinting

Signed-off-by: Jonathan Csanyi <jdc@csanyi.ca>

jcsanyi commented:
I added the type hinting you requested.

>My biggest issue with the patch is that, in order to follow consistency with pre-existing code, setters/getters are used to enable this behaviour, instead of relying on the Reader\Ini constructor.

I agree - I was thinking I should add constructor support for all 3 options as a follow-up PR.

weierophinney committed:
docs: one sentence per line

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove annotations where types work, update indentation

- Removes annotations where typehints exist
- Use `self`, not class name as a return type
- Indent here/nowdocs (available since PHP 7.3)

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
